go-controller/pkg/ovn/port_cache.go

```diff
@@ -0,0 +1,90 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/klog"
+)
+
+type portCache struct {
+	sync.RWMutex
+
+	counter  uint64
+	stopChan <-chan struct{}
+	cache    map[string]*lpInfo
+}
+
+type lpInfo struct {
+	name          string
+	uuid          string
+	logicalSwitch string
+	ip            net.IP
+	mac           net.HardwareAddr
+	// dead indicates that this object is scheduled to be removed
+	dead uint64
+}
+
+func newPortCache(stopChan <-chan struct{}) *portCache {
+	return &portCache{
+		stopChan: stopChan,
+		cache:    make(map[string]*lpInfo),
+	}
+}
+
+func (c *portCache) get(logicalPort string) (*lpInfo, error) {
+	c.RLock()
+	defer c.RUnlock()
+	if info, ok := c.cache[logicalPort]; ok {
+		return info, nil
+	}
+	return nil, fmt.Errorf("logical port %s not found in cache", logicalPort)
+}
+
+func (c *portCache) add(logicalSwitch, logicalPort, uuid string, mac net.HardwareAddr, ip net.IP) *lpInfo {
+	c.Lock()
+	defer c.Unlock()
+	portInfo := &lpInfo{
+		logicalSwitch: logicalSwitch,
+		name:          logicalPort,
+		uuid:          uuid,
+		ip:            ip,
+		mac:           mac,
+	}
+	c.cache[logicalPort] = portInfo
+	return portInfo
+}
+
+func (c *portCache) remove(logicalPort string) {
+	c.Lock()
+	defer c.Unlock()
+	info, ok := c.cache[logicalPort]
+	if !ok || info.dead > 0 {
+		return
+	}
+	info.dead = atomic.AddUint64(&c.counter, 1)
+
+	// Removal must be deferred since, for example, NetworkPolicy pod handlers
+	// may run after the main pod handler and look for items in the cache
+	// on delete events.
+	go func(stamp uint64) {
+		select {
+		case <-time.After(time.Minute):
+			c.Lock()
+			defer c.Unlock()
+			// remove the port info if the stamp matches what this
+			// goroutine expects. This makes sure that we don't
+			// prematurely remove a pod that was deleted and re-added
+			// before the timer expires.
+			if info, ok := c.cache[logicalPort]; ok && info.dead == stamp {
+				klog.Warningf("##### deleting port %s", logicalPort)
```

Comment:
it would be nice to have some debug logging around when add, remove updates the cache. For separate goroutines on a timer like this, it could become a real pain to debug down the road if the port cache is messed up and we have no idea what the state of it was

Why is this warning here?

Reply:
@trozet this was just a commit to test out whether the behavior I changed actually worked :) I intend to remove this commit and re-push

Reply:
@trozet added logging.